rust/src/operations/transaction/conflict_checker.rs

```diff
@@ -0,0 +1,862 @@
+//! Helper module to check if a transaction can be committed in case of conflicting commits.
+use std::collections::HashSet;
+use std::io::{BufRead, BufReader, Cursor};
+
+use object_store::ObjectStore;
+
+use super::CommitInfo;
+use crate::action::{Action, Add, DeltaOperation, MetaData, Protocol, Remove};
+use crate::delta_config::IsolationLevel;
+use crate::storage::commit_uri_from_version;
+use crate::{table_state::DeltaTableState, DeltaDataTypeVersion, DeltaResult, DeltaTableError};
+
+#[cfg(feature = "datafusion")]
+use super::state::AddContainer;
+#[cfg(feature = "datafusion")]
+use datafusion_expr::Expr;
+#[cfg(feature = "datafusion")]
+use itertools::Either;
+
+/// Exceptions raised during commit conflict resolution
+#[derive(thiserror::Error, Debug)]
+pub enum CommitConflictError {
+    /// This exception occurs when a concurrent operation adds files in the same partition
+    /// (or anywhere in an un-partitioned table) that your operation reads. The file additions
+    /// can be caused by INSERT, DELETE, UPDATE, or MERGE operations.
+    #[error("Concurrent append failed.")]
```

Comment:
ConcurrentAppend is what went wrong, not what you are trying to do, right?

```suggestion
    #[error("Commit failed: a concurrent transactions added new files.\nHelp: This transaction's query must be rerun to include the new data. Also, if you don't care to require this check to pass in the future, the isolation level can be set to Snapshot Isolation.")]
```